tests/testtext.c

```diff
@@ -0,0 +1,214 @@
+#ifdef WIN32
+#ifndef __cplusplus
+#error Please compile with a C++ compiler.
+#endif
+#endif
+
+#if defined(USE_WINDOWS_GDIPLUS)
+#include <Windows.h>
+#include <GdiPlus.h>
+
+#pragma comment(lib, "gdiplus")
+#else
+#include <GdiPlusFlat.h>
+#endif
+
+#if defined(USE_WINDOWS_GDIPLUS)
+using namespace Gdiplus;
+using namespace DllExports;
+#endif
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include "testhelpers.h"
+#include "../config.h"
+
+#define ok(expected, ...) if (!(expected)) { printf(__VA_ARGS__); assert(expected); }
+#define expect(expected, got) ok((got) == (expected), "Expected %d, got %d\n", (INT)(expected), (INT)(got))
+#define expectf_(expected, got, precision) ok(fabs((expected) - (got)) <= (precision), "Expected %f, got %f\n", (expected), (got))
+#define expectf(expected, got) expectf_((expected), (got), 0.001)
+#define TODO if (0)
+#define set_rect_empty(r) (r)->X = (r)->Y = (r)->Width = (r)->Height = 0
+
+#ifdef USE_PANGO_RENDERING
+
+static void test_measure_string(void)
+{
+	GpStringFormat *format;
+	GpImage *image;
+	GpGraphics *graphics;
+	GpFontFamily *family;
+	GpFont *font;
+	GpStatus status;
+	GpRectF rect, bounds, saved_bounds;
+	const WCHAR teststring1[] = { 'M', '\n', '\n', 'M', 0 };
+	int glyphs;
+	int lines;
+
+	status = GdipCreateStringFormat (0, 0, &format);
+	expect (Ok, status);
+	status = GdipGetGenericFontFamilySansSerif (&family);
+	expect (Ok, status);
+	status = GdipCreateFont (family, 10, FontStyleRegular, UnitPixel, &font);
+	expect (Ok, status);
+	status = GdipCreateBitmapFromScan0 (400, 400, 0, PixelFormat32bppRGB, NULL, &image);
+	expect (Ok, status);
+	status = GdipGetImageGraphicsContext (image, &graphics);
+	expect (Ok, status);
+	ok (graphics != NULL, "Expected graphics to be initialized\n");
+
+	rect.X = 5.0;
+	rect.Y = 5.0;
+	rect.Width = 200.0;
+	rect.Height = 200.0;
+	set_rect_empty (&bounds);
+	status = GdipMeasureString (graphics, teststring1, 1, font, &rect, format, &bounds, &glyphs, &lines);
+	expect (Ok, status);
+	expectf (5.0, bounds.X);
+	expectf (5.0, bounds.Y);
+	expect (1, glyphs);
+	expect (1, lines);
+
+	// New lines count as fitted code points
+	set_rect_empty (&bounds);
+	status = GdipMeasureString (graphics, teststring1, 2, font, &rect, format, &bounds, &glyphs, &lines);
+	expect (Ok, status);
+	expect (2, glyphs);
+	expect (1, lines);
+	saved_bounds = bounds;
+
+	// Multiple new lines are counted as code points and lines
+	set_rect_empty (&bounds);
+	status = GdipMeasureString (graphics, teststring1, 3, font, &rect, format, &bounds, &glyphs, &lines);
+	expect (Ok, status);
+	TODO expect (3, glyphs);
+	expect (2, lines);
+
+	// ...and lastly check the correct results for adding another character past the last new line
+	set_rect_empty (&bounds);
+	status = GdipMeasureString (graphics, teststring1, 4, font, &rect, format, &bounds, &glyphs, &lines);
+	expect (Ok, status);
+	TODO expect (4, glyphs);
```

Comment:
Do we still need these todos?

Reply:
(I’m assuming that these are for wine instead of for libgdiplus)

Maybe we can also use the #if USE_WINDOWS_GDIPLUS macro to be clearer about this

Reply:
These test are not from Wine (unlike the other PR). Yes, we need them, unfortunately.

Reply:
I can wrap the `TODO` macro with `#if USE_WINDOWS_GDIPLUS`. I'd prefer that over having `#if` all over the place.